packages/jest-mock/src/index.ts

```diff
@@ -912,26 +924,26 @@ class ModuleMockerClass {
     return fn;
   }
 
-  spyOn<T extends {}, M extends keyof T>(
+  spyOn<T extends {}, M extends NonFunctionPropertyNames<T>>(
     object: T,
     methodName: M,
     accessType: 'get',
   ): SpyInstance<T[M], []>;
 
-  spyOn<T extends {}, M extends keyof T>(
+  spyOn<T extends {}, M extends NonFunctionPropertyNames<T>>(
     object: T,
     methodName: M,
     accessType: 'set',
   ): SpyInstance<void, [T[M]]>;
 
-  spyOn<T extends {}, M extends keyof T>(
+  spyOn<T extends {}, M extends FunctionPropertyNames<T>>(
     object: T,
     methodName: M,
   ): T[M] extends (...args: any[]) => any
     ? SpyInstance<ReturnType<T[M]>, ArgsType<T[M]>>
     : never;
 
-  spyOn<T extends {}, M extends keyof T>(
+  spyOn<T extends {}, M extends NonFunctionPropertyNames<T>>(
```

Comment:
heh, not this is not entirely true because this generic `spyOn` can accept function _and_ non-functional property names, but TS seems to buy it. 

I don't really understand why a function even needs type definitions when it's overloaded like this. Generated types are only for the definitions as well, so that seems really counterintuitive, not to mention it causes headaches on "how to fit all the typings in this def and make TS happy, even though it may be a lie" 😅

Reply:
heh, yeah. we might want to add actual tests on the types at some point (e.g. https://www.npmjs.com/package/tsd-check or https://www.npmjs.com/package/dts-jest)